### PR TITLE
MaterializeMySQL: Attempt to reconnect to MySQL if the connection is lost

### PR DESCRIPTION
**Pull request by havardk**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
MaterializeMySQL: Attempt to reconnect to MySQL if the connection is lost

Detailed description / Documentation draft:
If the connection to MySQL is lost (e.g. if MySQL is restarted) attempt to reconnect by default.  Also, if MySQL is not available when an existing database is loaded when ClickHouse is started, don't allow tables to be queried unless the `allows_query_when_mysql_lost` setting is enabled.

### CODE REVIEW

File: src/Databases/MySQL/MaterializeMetadata.cpp [hunk 1]
```diff
@@ -239,9 +237,16 @@ MaterializeMetadata::MaterializeMetadata(
         assertString("\nData Version:\t", in);
         readIntText(data_version, in);
 
-        if (checkBinlogFileExists(connection))
-            return;
     }
+}
+
+void MaterializeMetadata::startReplication(
```
Comment (stigsb) [inline src/Databases/MySQL/MaterializeMetadata.cpp:243]: Would it be a good idea to maintain a "running" state in MaterializeMetadata, and add a "stopReplication" method?

Comment (havardk) [reply src/Databases/MySQL/MaterializeMetadata.cpp:243]: Perhaps.  It is not necessary now though, so I think that should be added when the need arises.

**Comment by havardk** [comment]
Hi,

It took some time to get this PR in shape, but I think it is OK now.  The test failures look unrelated, though I don't know about the Yandex internal ones.  Should I try to run the tests again?

**Comment by tavplubix** [comment]
Hi, you can just ignore Yandex synchronization check failure, because it's only for Yandex employees.
Integration tests (asan) - test_distributed_ddl - broken in master
Performance - flaky

**Review by tavplubix — APPROVED** [review]

**Pull request merged**